
<!DOCTYPE html>
<html>
<head><meta name="generator" content="Hexo 3.8.0">
    <meta charset="UTF-8">
    <title>Date with Bioawk | Universe Biology Girls and Boys</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <meta name="keywords" content>
    <meta name="description" content="原创：老板，来一打TPU
废话这是生平第一篇blog，因此难免要说点废话，所以前戏，啊不对，前言就叫废话吧。git+hexo搭建的平台杠杠的，感谢师弟MingLian小大佬成功带入写blog的坑。第一篇就拿生信李恒大神的神作bioawk镇楼吧！**
bioawk简介除了鼎鼎大名的BWA、samtools等，李恒大神应群众呼声又做了出于awk又胜于awk的bioawk供广大生信猿们玩耍。bioawk是用C写的，用法和awk基本一样。
安装分步安装1234567891011121314151617    $ sudo apt-get install byacc git    #安装git    $ git clone git...">
    
        <link rel="icon" href="/favicon.ico">
    
    
        
            <link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css">
        
            <link rel="stylesheet" href="/css/stage.css">
        
            <link rel="stylesheet" href="/css/avatar-bg.css">
        
    
</head>
</html>
<body>
<header id="header">
    <div class="menu">
        <i class="fa fa-bars"></i>
    </div>
    <div class="header-main">
        <h1><a href="/">Universe Biology Girls and Boys</a></h1>
    </div>
    <div id="nav">
        <div class="nav-img" id="nav-img"></div>
        <div class="sentences">
            我跑的胶，叠起来可以绕地球一整圈
	</div>
    </div>
</header>

<div id="content-outer">
    <div id="content-inner">
        <div class="clearfix">
    <article id="post">
        <h1>Date with Bioawk</h1>
        <div class="create">
            <span>Created</span>
            
                <time datetime="2019-03-16T08:39:23.000Z">
                    2019-03-16
                </time>
            
            <ul class="article-category-list"><li class="article-category-list-item"><a class="article-category-list-link" href="/categories/Software/">Software</a></li></ul>
        </div>
        <p><strong>原创：老板，来一打TPU</strong></p>
<h1 id="废话"><a href="#废话" class="headerlink" title="废话"></a>废话</h1><p>这是生平第一篇blog，因此难免要说点废话，所以前戏，啊不对，前言就叫废话吧。git+hexo搭建的平台杠杠的，感谢师弟MingLian小大佬成功带入<br>写blog的坑。第一篇就拿生信李恒大神的神作bioawk镇楼吧！**<br><a id="more"></a></p>
<h1 id="bioawk简介"><a href="#bioawk简介" class="headerlink" title="bioawk简介"></a>bioawk简介</h1><p>除了鼎鼎大名的BWA、samtools等，李恒大神应群众呼声又做了出于awk又胜于awk的bioawk<br>供广大生信猿们玩耍。bioawk是用C写的，用法和awk基本一样。</p>
<h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><h3 id="分步安装"><a href="#分步安装" class="headerlink" title="分步安装"></a>分步安装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">    $ sudo apt-get install byacc git    <span class="comment">#安装git</span></span><br><span class="line">    $ git <span class="built_in">clone</span> git://github.com/lh3/bioawk.git  <span class="comment">#默认安装在当前目录</span></span><br><span class="line">    git <span class="built_in">clone</span> git://github.com/lh3/bioawk.git <span class="variable">$&#123;path&#125;</span>/bioawk <span class="comment">#若加path则必须为空，因此可在想要安装的目录下起一个新名称，git会clone到其下</span></span><br><span class="line">    $ <span class="built_in">cd</span> bioawk</span><br><span class="line">    $ make</span><br><span class="line">    $ <span class="built_in">echo</span> <span class="string">"export PATH=<span class="variable">$&#123;path&#125;</span>/bioawk/:\$&#123;PATH&#125;"</span> &gt;&gt; ~/.bashrc</span><br><span class="line">    $ .  ~/.bashrc</span><br><span class="line"></span><br><span class="line"><span class="comment">### 一步安装脚本</span></span><br><span class="line">    <span class="comment">#!/bin/bash</span></span><br><span class="line">    <span class="built_in">read</span> -p  <span class="string">'Input installed  path: '</span> tmp_path</span><br><span class="line">    path=<span class="variable">$&#123;tmp_path/\//&#125;</span></span><br><span class="line">    <span class="keyword">if</span> [ ! -d <span class="string">"<span class="variable">$&#123;path&#125;</span>"</span> ];<span class="keyword">then</span> <span class="built_in">echo</span> <span class="string">"No such file:<span class="variable">$&#123;path&#125;</span>"</span> &amp;&amp; <span class="built_in">exit</span> ;<span class="keyword">fi</span></span><br><span class="line">    git <span class="built_in">clone</span> git://github.com/lh3/bioawk.git <span class="variable">$&#123;path&#125;</span>/bioawk &amp;&amp; \</span><br><span class="line">    <span class="built_in">cd</span> <span class="variable">$&#123;path&#125;</span>/bioawk &amp;&amp; make &amp;&amp; \</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">"export PATH=<span class="variable">$&#123;path&#125;</span>/bioawk/:\$&#123;PATH&#125;"</span> &gt;&gt; ~/.bashrc  &amp;&amp; .  ~/.bashrc &amp;&amp; \</span><br><span class="line">    <span class="built_in">echo</span> <span class="string">'Successfully install bioawk!'</span> || <span class="built_in">echo</span> <span class="string">'Failed install bioawk'</span></span><br></pre></td></tr></table></figure>
<h2 id="使用"><a href="#使用" class="headerlink" title="使用"></a>使用</h2><p>bioawk基本思想是把组成不同类型的文件（sam、bam、fasta、fastq、vcf etc）的基本元<br>素封装成变量，直接调用即可。</p>
<h3 id="语法"><a href="#语法" class="headerlink" title="语法"></a>语法</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">usage: bioawk [-F fs] [-v var=value] [-c fmt] [-tH] [-f progfile | <span class="string">'prog'</span>] [file ...]</span><br><span class="line"></span><br><span class="line">-c 支持输入文件格式，查看帮助：</span><br><span class="line"></span><br><span class="line">    bioawk -c -h</span><br><span class="line"></span><br><span class="line">bed:</span><br><span class="line">        1:chrom 2:start 3:end 4:name 5:score 6:strand 7:thickstart 8:thickend 9:rgb 10:blockcount 11:blocksizes 12:blockstarts</span><br><span class="line">sam:</span><br><span class="line">        1:qname 2:flag 3:rname 4:pos 5:mapq 6:cigar 7:rnext 8:pnext 9:tlen 10:seq 11:qual</span><br><span class="line">vcf:</span><br><span class="line">        1:chrom 2:pos 3:id 4:ref 5:alt 6:qual 7:filter 8:info</span><br><span class="line">gff:</span><br><span class="line">        1:seqname 2:<span class="built_in">source</span> 3:feature 4:start 5:end 6:score 7:strand 8:frame 9:attribute</span><br><span class="line">fastx:</span><br><span class="line">        1:name 2:seq 3:qual 4:comment</span><br></pre></td></tr></table></figure>
<p><strong>上面出现的名称即可引用其变量</strong></p>
<h3 id="实例"><a href="#实例" class="headerlink" title="实例"></a>实例</h3><ul>
<li>打印fasta序列ID、序列、长度、GC含量<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bioawk -cfastx <span class="string">'&#123;print "ID: "$name"\tlength: "length($seq)"\tGC: "gc($seq)"\t"$seq&#125;'</span> demo.fa</span><br></pre></td></tr></table></figure>
</li>
</ul>
<p>结果:<br>    ID: g1  length: 18      GC: 0.722222    atcccccccccccccttt<br>    ID: g2  length: 26      GC: 0.0769231   cattatatcttatttttttttttttt<br>    ID: g3  length: 48      GC: 0.229167    acccccccccctttttttttttttcatttttttttttttttttttttt<br>原始文件：<br>    &gt;g1<br>    atcccccccccccccttt<br>    &gt;g2 enzyme<br>    cattatatcttatttttttttttttt<br>    &gt;g3<br>    accccccccccttttttttttttt<br>    catttttttttttttttttttttt<br>注意：<br><strong>可以看到bioawk在提取ID的时候只提取了空格分隔后的第一个字段，默认输出域分割符是\t</strong></p>
<ul>
<li><p>输出反向互补序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bioawk -cfastx <span class="string">'&#123;print $name,$seq,revcomp($seq)&#125;'</span> demo.fa</span><br></pre></td></tr></table></figure>
<p>结果：<br>  g1      atcccccccccccccttt      aaagggggggggggggat<br>  g2      cattatatcttatttttttttttttt      aaaaaaaaaaaaaataagatataatg<br>  g3      acccccccccctttttttttttttcatttttttttttttttttttttt        aaaaaaaaaaaaaaaaaaaaaatgaaaaaaaaaaaaaggggggggggt</p>
</li>
<li><p>根据序列ID提取序列，这个是最爱的功能</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bioawk -cfastx <span class="string">'BEGIN&#123;while((getline k &lt;"id.txt")&gt;0)i[k]=1&#125;&#123;if(i[$name])print "&gt;"$name"\n"$seq&#125;'</span> demo.fa</span><br></pre></td></tr></table></figure>
</li>
</ul>
<p>id.txt:<br>    g1<br>    g2<br>    g2 enzyme<br>    &gt;g3</p>
<p>结果：<br>    &gt;g1<br>    atcccccccccccccttt<br>    &gt;g2<br>    cattatatcttatttttttttttttt<br>注意：id.txt 需要去掉&gt;</p>
<ul>
<li>当然还有其他的花样，大家自己去探索吧</li>
</ul>
<h1 id="Reference"><a href="#Reference" class="headerlink" title="Reference"></a>Reference</h1><p><a href="https://github.com/lh3/bioawk" target="_blank" rel="noopener">https://github.com/lh3/bioawk</a></p>

        <div>
            <ul class="tags-category-list"><li class="tags-category-list-item"><a class="tags-category-list-link" href="/tags/bioawk/">bioawk</a></li><li class="tags-category-list-item"><a class="tags-category-list-link" href="/tags/bioinformatics/">bioinformatics</a></li></ul>
        </div>
        <div class="bottom-line"></div>
        
    <nav id="article-nav">
        
            <a href="/2019/03/19/paper-submit-experience/" id="article-nav-newer" class="article-nav-link-wrap">
        <span class="article-nav-title">
            
                浅谈科研论文发表与投稿-下篇
            
        </span>
                <strong class="article-nav-caption">&gt;</strong>
            </a>
        
        
            <a href="/2019/03/16/MARKDOWN-s-KFC/" id="article-nav-older" class="article-nav-link-wrap">
                <strong class="article-nav-caption">&lt;</strong>
                <span class="article-nav-title">
                    
                        MARKDOWN&#39;s KFC
                </span>
                
            </a>
        
    </nav>


        
            <div id="stageComment" class="comment" style="margin-top: 27px;"></div>
<script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script>
<script src="//unpkg.com/valine/dist/Valine.min.js"></script>
<script>
  var notify = 'false' == true ? true : false;
  var verify = 'false' == true ? true : false;
  window.onload = function() {
    new Valine({
      el: '#stageComment',
      notify: notify,
      verify: verify,
      app_id: "aok3pCDe5ao2IIhmO7Cwjmgi-gzGzoHsz",
      app_key: "QT8zl1KLRblAYwCsuCcj6JAz",
      placeholder: "小姐姐、小哥哥，快来夸我啊!",
      avatar:"mm"
    });
  }
</script>

        
    </article>
    <div id="toc">
        
            <h2>Table of Contents</h2>
            <ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#废话"><span class="toc-number">1.</span> <span class="toc-text">废话</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#bioawk简介"><span class="toc-number">2.</span> <span class="toc-text">bioawk简介</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#安装"><span class="toc-number">2.1.</span> <span class="toc-text">安装</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#分步安装"><span class="toc-number">2.1.1.</span> <span class="toc-text">分步安装</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#使用"><span class="toc-number">2.2.</span> <span class="toc-text">使用</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#语法"><span class="toc-number">2.2.1.</span> <span class="toc-text">语法</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#实例"><span class="toc-number">2.2.2.</span> <span class="toc-text">实例</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Reference"><span class="toc-number">3.</span> <span class="toc-text">Reference</span></a></li></ol>
        
    </div>
</div>

    </div>
</div>
<footer id="footer">
    <div id="copyright">&copy; Uranusyt  2019</div>
    <div id="theme">
        Powered by <a href="http://hexo.io">Hexo</a>. Theme by <a href="https://github.com/markyong/hexo-theme-stage">Stage</a>
    </div>
</footer>
<script src="/lib/js/waterrippleeffect.min.js"></script>
<script src="/js/header-bg.main.js"></script>

    <script src="/lib/js/smooth-scroll.min.js"></script>
    <script src="/js/toc.main.js"></script>
    <script async src="//dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span id="busuanzi_container_site_uv"> 
  本站访客数<span id="busuanzi_value_site_uv"></span>人次
</span>
    <span id="busuanzi_container_page_pv">
   本文总阅读量<span id="busuanzi_value_page_pv"></span>次
</span>

<script src="/live2dw/lib/L2Dwidget.min.js?0c58a1486de42ac6cc1c59c7d98ae887"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"log":false,"model":{"jsonPath":"/live2dw/assets/z16.model.json"},"display":{"position":"right","width":150,"height":300},"mobile":{"show":true}});</script></body>
</html>
